File: week2/Ariling/boj-2636-치즈/main.js
const readline = require("readline");
const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

function solution(input, total) {
  let pre_cheese = total;
  let day = 0;
  while (total > 0) {
    day++;
    pre_cheese = total;
    let check = Array(arrInfo[0])
      .fill()
      .map(() => Array(arrInfo[1]).fill(false));
    bfs(0, 0, check);
    bfs(0, input[0].length - 1, check);
    bfs(input.length - 1, 0, check);
    bfs(input.length - 1, input[0].length - 1, check);
    for (let i = 0; i < input.length; i++) {
      for (let j = 0; j < input[i].length; j++) {
        if (input[i][j] === 1 && check[i][j] == true) {
          input[i][j] = 0;
          total--;
        }
      }
    }
  }
  console.log(day);
  console.log(pre_cheese);
  return;
}

function bfs(x, y, check) {
  let bfs_arr = [[x, y]];
  check[x][y] = true;
  var dr = [-1, 1, 0, 0];
  var dc = [0, 0, -1, 1];
  while (bfs_arr.length > 0) {
    [r, c] = bfs_arr.shift();
    for (let i = 0; i < 4; i++) {
      newR = r + dr[i];
      newC = c + dc[i];
      if (
        0 <= newR &&
        newR < arrInfo[0] &&
        0 <= newC &&
        newC < arrInfo[1] &&
        check[newR][newC] == false
      ) {
        if (arr[newR][newC] == 0) {
          bfs_arr.push([newR, newC]);
        }
        check[newR][newC] = true;
      }
    }
  }
}

let arrInfo = [];
let arr = [];

rl.on("line", (line) => {
  if (arrInfo.length == 0) {
    arrInfo = line.split(" ").map(Number);
  } else {
    arr.push(line.split(" ").map(Number));
    if (arr.length === arrInfo[0]) {
      rl.close();
    }
  }
}).on("close", () => {
  let count = 0;
  for (let i = 0; i < arr.length; i++) {
    for (let j = 0; j < arr[i].length; j++) {
      if (arr[i][j] === 1) count++;
    }
  }
  solution(arr, count);
  process.exit();
});

File: week2/Ariling/boj-2636-치즈/main.py
import sys
from collections import deque
input = sys.stdin.readline

arr_info = list(map(int, input().split()))
arr = [list(map(int, input().split())) for _ in range(arr_info[0])]
total_cheese = 0
answer = 0
time = 0

for i in range(1, arr_info[0]-1):
    for j in range(1, arr_info[1]-1):
        if arr[i][j] == 1:
            total_cheese += 1
dr = [-1, 1, 0, 0]
dc = [0, 0, -1, 1]
def bfs(x, y):
    global checked
    q = deque([(x, y)])
    checked[x][y] = True
    while q:
        r, c = q.popleft()
        for i in range(4):
            newR = r + dr[i]
            newC = c + dc[i]
            if 0 <= newR < arr_info[0] and 0 <= newC < arr_info[1] and not checked[newR][newC]:
                if arr[newR][newC] == 1:
                    checked[newR][newC] = True
                    continue
                else:
                    checked[newR][newC] = True
                    q.append((newR, newC))

while total_cheese > 0:
    # 전의 치즈 갯수를 담기 위함
    answer = total_cheese
    checked = [[False] * arr_info[1] for _ in range(arr_info[0])]
    # 아? 가장자리만 하면 되는구나?
    bfs(0, 0)
    bfs(arr_info[0]-1, 0)
    bfs(0, arr_info[1]-1)
    bfs(arr_info[0]-1, arr_info[1]-1)
    disappear_cheese = 0
    for i in range(1, arr_info[0] - 1):
        for j in range(1, arr_info[1] - 1):
            if arr[i][j] == 1 and checked[i][j]:
                disappear_cheese += 1
                arr[i][j] = 0
    total_cheese -= disappear_cheese
    time += 1

print(time)
print(answer)


